hyperspy/_signals/hologram_image.py

```
@@ -0,0 +1,329 @@
+                              'The reconstruction will be performed with the first value.')
+                sb_size_temp = np.float64(sb_size.inav[0].data)
+            else:  # sb_position navdim=0, therefore map function should not iterate it:
+                sb_size_temp = np.float64(sb_size.data)
+        else:
+            sb_size_temp = sb_size.deepcopy()
+        #
+
+        # Standard edge smoothness of sideband aperture 5% of sb_size
+        if sb_smoothness is None:
+            sb_smoothness = sb_size * 0.05
+        else:
+            if not isinstance(sb_smoothness, BaseSignal):
+                if isinstance(sb_smoothness, np.ndarray) and sb_smoothness.size > 1:
+                    sb_smoothness = BaseSignal(sb_smoothness).T
+                else:
+                    sb_smoothness = BaseSignal(sb_smoothness)
+
+        if sb_smoothness.axes_manager.navigation_size != self.axes_manager.navigation_size:
+            if sb_smoothness.axes_manager.navigation_size:
+                warnings.warn('Sideband smoothness dimensions do not match neither reference nor hologram dimensions.'
+                              'The reconstruction will be performed with the first value.')
+                sb_smoothness_temp = np.float64(sb_smoothness.inav[0].data)
+            else:  # sb_position navdim=0, therefore map function should not iterate it:
+                sb_smoothness_temp = np.float64(sb_smoothness.data)
+        else:
+            sb_smoothness_temp = sb_smoothness.deepcopy()
+
+        # Convert sideband size from 1/nm or mrad to pixels
+        if sb_unit == 'nm':
+            sb_size_temp = sb_size_temp / np.mean(self.f_sampling)
+            sb_smoothness = sb_smoothness / np.mean(self.f_sampling)
+        elif sb_unit == 'mrad':
+            try:
+                ht = self.metadata.Acquisition_instrument.TEM.beam_energy
+            except AttributeError:
+                ht = int(input('Enter beam energy in kV: '))
```

I would rather raise an exception explaining that the information is missing and that it can be provided using `set_microscope_parameters`.

---

Two reasons why I would leave it as it is at the moment:
1) It is faster for a user.
2) `set_microscope_parameters` doesn't seem to be a method available for all signals. In particular for me it seem it only works for `EDS_SEM`, even in `EELS` I get `NotImplementedError`. So it should be implemented in `HologramImage` first.

---

That's a good point. We should probably have a `TEM` superclass from which all `TEM` signals inherit that includes `set_microscope_parameters`. It should work for EELS though. If you get a `NotImplementedError` is probably because you didn't initialize the backend before running the function e.g. `%matplotlib qt4`.

I don't think that it is a good idea to ask for input interactively. For example, [HyperSpyUI](http://hyperspy.org/hyperspyUI/) would struggle with it. If it raises an error, the user can then use `set_microscope_parameters` that includes a UI for HyperSpyUI. This is what we currently do for other methods that require missing, [here is an example](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/_signals/eels.py#L1050).

---

I'm not sure what is wrong there with the `set_microsocpe_parameters` but here is what I've got either switching to qt4 backend or initialising my standard OS X backend:

> In[3]: %matplotlib qt4
In[4]: aa = hs.signals.EELSSpectrum(np.reshape(np.arange(16), (2, 2, 2, 2)))
In[5]: aa
Out[5]: <EELSSpectrum, title: , dimensions: (2, 2, 2|2)>
In[6]: aa.set_microscope_parameters(beam_energy=300)
Traceback (most recent call last):
  File "/Users/vadim/anaconda/lib/python3.5/site-packages/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-2477a9aea25a>", line 1, in <module>
    aa.set_microscope_parameters(beam_energy=300)
  File "/Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/_signals/eels.py", line 833, in set_microscope_parameters
    self._are_microscope_parameters_missing()
  File "/Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/_signals/eels.py", line 793, in _are_microscope_parameters_missing
    is_ok = messagesui.information(par_str)
  File "/Users/vadim/Work_Recent/Prog/_Py/hyperspy/hyperspy/gui/messages.py", line 75, in information
    message.edit_traits()
  File "/Users/vadim/anaconda/lib/python3.5/site-packages/traits/has_traits.py", line 1859, in edit_traits
    handler, id, scrollable, args )
  File "/Users/vadim/anaconda/lib/python3.5/site-packages/traitsui/view.py", line 433, in ui
    ui.ui( parent, kind )
  File "/Users/vadim/anaconda/lib/python3.5/site-packages/traitsui/ui.py", line 222, in ui
    self.rebuild( self, parent )
  File "/Users/vadim/anaconda/lib/python3.5/site-packages/traitsui/toolkit.py", line 226, in ui_modal
    raise NotImplementedError
NotImplementedError
In[7]: aa.set_signal_type('EDS_SEM')
In[8]: aa
Out[8]: <EDSSEMSpectrum, title: , dimensions: (2, 2, 2|2)>
In[9]: aa.set_microscope_parameters(beam_energy=300)
In[10]: aa.metadata.Acquisition_instrument.SEM.beam_energy
Out[10]: 300

---

So what shall I do? I can simply copy a `set_microscope_parameters` method from `EDS_SEM` to `HologramImage` or can try creating `TEM` class and let `HologramImage` to inherit from it if there is a consensus about this...

---

Shall I post the above as an issue or is it a problem of my particular installation?

---

The exception is raised by traitsui, so, almost certainly it has to do not initializing the backend properly. Could try restarting the kernel and import hyperspy as follows (the order is important):

```python

%matplotlib qt4

import hyperspy.api as hs

```

Regarding `set_microscope_parameters`, for the moment you could simply copy the one in the EELS or SEM class and adapt it to your needs. We'll have to think on a more general solution to the problem, but it'll be probably better to leave this for a different PR.

---

This way it works, though with qt4 only...
Will copy one from `EDS_SEM` than.